### ISSUE TITLE
"Not-equal" pushdown filter removes too many documents

### ISSUE THREAD

**Issue by szczeles** [elastic/elasticsearch-hadoop#1133]
### What kind an issue is this?

- [x] Bug report

### Issue description

When using "not-equal" or "not-like" filters in spark with pushdown enabled, too many records are filtered out.

### Steps to reproduce

Test index is created with the following code:

```
spark.createDataFrame([
    (1, 'hello-world'), 
    (2, 'hello-kitty'), 
    (3, 'something-else'),
    (4, 'hellohello')], ['id', 'message']) \
    .write.format("org.elasticsearch.spark.sql").save('test-index/type')
```
When queried with pushdown **disabled**  and request for records with message not equal to "hello-world" the results are valid (as whole dataset is downloaded to spark context and spark itself removes the unwanted row):

```
spark.read.format("org.elasticsearch.spark.sql") \
    .option('pushdown', 'false') \
    .load('test-index/type').where(col('message') != 'hello-world').show()
+---+--------------+
| id|       message|
+---+--------------+
|  4|    hellohello|
|  3|something-else|
|  2|   hello-kitty|
+---+--------------+
```

But when using the same code with pushdown **enabled** (the default behavior), "hello-kitty" row is removed, but "hellohello" is still there...

```
spark.read.format("org.elasticsearch.spark.sql") \
    .load('test-index/type').where(col('message') != 'hello-world').show()
+---+--------------+
| id|       message|
+---+--------------+
|  3|something-else|
|  4|    hellohello|
+---+--------------+
```

If querying with not-like, results are the same:

```
spark.read.format("org.elasticsearch.spark.sql") \
    .load('test-index/type').where(~col('message').like('hello-world')).show()
+---+--------------+
| id|       message|
+---+--------------+
|  3|something-else|
|  4|    hellohello|
+---+--------------+
```

### Version Info

OS:         :  Ubuntu 16.04
JVM         :  1.8.0, Oracle, update 161
Hadoop/Spark:  Spark 2.3.0, Hadoop 2.7
ES-Hadoop   : [6.2.3 latest daily build](https://oss.sonatype.org/content/repositories/snapshots/org/elasticsearch/elasticsearch-hadoop/6.2.3-SNAPSHOT/elasticsearch-hadoop-6.2.3-20180320.011313-62.jar)
ES          :  6.2.3

**Comment by jbaiera** [comment]
This doesn't look like a bug so much as it's more of a misunderstanding of how the mapping works in Elasticsearch. If I create an index with the provided test data:

```bash
$> curl -XPUT http://localhost:9500/playground/doc/_bulk -H 'Content-Type: application/json' -d '{"index": { "_id": "1" } }
{"id": 1, "message": "hello-world"}
{"index": { "_id": "2" } }
{"id": 2, "message": "hello-kitty"}
{"index": { "_id": "3" } }
{"id": 3, "message": "something-else"}
{"index": { "_id": "4" } }
{"id": 4, "message": "hellohello"}
'
```

... and then do a text analysis of "hello-world" through the "message" field:

```bash
$> curl -X POST http://localhost:9500/playground/_analyze -H 'Content-Type: application/json' -d '{
	"field": "message",
	"text": "hello-world"
}'
```
... then it returns that the field is being tokenized on the dash symbol:
```json
{
    "tokens": [
        {
            "token": "hello",
            "start_offset": 0,
            "end_offset": 5,
            "type": "<ALPHANUM>",
            "position": 0
        },
        {
            "token": "world",
            "start_offset": 6,
            "end_offset": 11,
            "type": "<ALPHANUM>",
            "position": 1
        }
    ]
}
```
The pushdown logic in Spark is going to generate something close to the following query:
```json
{
  "query": {
    "bool": {
      "must_not": [
        {"match": {"message": "hello-world"}}
      ]
    }
  }
}
```
... and that query's match text will be analyzed with the same analyzer as the message field by default. This will produce the "hello" and "world" tokens, of which the "hello" token will match "hello-world" and "hello-kitty".

If you are looking to use the not-logic in pushdown and are looking for exact matches, then I suggest supplying the mapping for the index ahead of time and using the "keyword" field type instead of text.

**Comment by szczeles** [comment]
Hi James, thank you for the response. Actually I already managed to get it to work without changing the mapping, I used `strict` option and now [the term is used instead of match](https://github.com/elastic/elasticsearch-hadoop/blob/master/spark/sql-20/src/main/scala/org/elasticsearch/spark/sql/DefaultSource.scala#L347). As a spark developer I was  suprised that the library doesn't work as usual spark source (where "not equal" is just "not equal"). but  I understand your motivation of setting default value of `strict` to `false`. 

Thanks again :-)
